### PR TITLE
Metadata: Implement storage/retrieval of level/profile visibility

### PR DESCRIPTION
**Pull request by Beyley**
This does not implement the functionality, but it means we are now storing all the relevant information to make that happen.

### CODE REVIEW

File: Refresh.GameServer/Database/GameDatabaseContext.Users.cs [hunk 1]
```diff
@@ -380,4 +386,20 @@ public void IncrementUserFilesizeQuota(GameUser user, int amount)
...
+        });
+    }
+    
+    public void SetLevelVisibility(GameUser user, Visibility visibility)
+    {
+        this._realm.Write(() =>
+        {
+            user.LevelVisibility = visibility;
+        });
+    }
```
Comment (jvyden) [inline Refresh.GameServer/Database/GameDatabaseContext.Users.cs:404]: These parameters technically aren't bundled in one request from the game, but to simplify the endpoint and to keep this class shorter I think it should be one function called `SetUserPrivacySettings`.

File: Refresh.GameServer/Endpoints/ApiV3/DataTypes/Request/ApiUpdateUserRequest.cs [hunk 2]
```diff
@@ -14,4 +16,7 @@ public class ApiUpdateUserRequest
     public bool? UnescapeXmlSequences { get; set; }
     
     public string? EmailAddress { get; set; }
+    
+    public Visibility? LevelVisibility { get; set; }
+    public Visibility? ProfileVisibility { get; set; }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/ApiV3/DataTypes/Request/ApiUpdateUserRequest.cs:21]: New properties in this class should be handled in the corresponding function in `GameDatabaseContext`. The function in question is the `ApiUpdateUserRequest` overload of `UpdateUserData`.

This makes it so these properties can actually be updated from the API - we removed the reflection-based updating a while ago.

File: Refresh.GameServer/Endpoints/Game/Handshake/MetadataEndpoints.cs [hunk 3]
```diff
@@ -18,15 +18,31 @@ public class MetadataEndpoints : EndpointGroup
...
+        {
+            LevelVisibility = user.LevelVisibility,
+            ProfileVisibility = user.ProfileVisibility,
+        };
     }
     
     [GameEndpoint("privacySettings", ContentType.Xml, HttpMethods.Post)]
-    public PrivacySettings SetPrivacySettings(RequestContext context)
+    [MinimumRole(GameUserRole.Restricted)]
+    [DebugRequestBody]
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Handshake/MetadataEndpoints.cs:32]: `[DebugRequestBody]` Should be nuked

**Review by jvyden — CHANGES_REQUESTED** [review]

File: Refresh.GameServer/Endpoints/ApiV3/DataTypes/Request/ApiUpdateUserRequest.cs (hunk 2 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/ApiV3/DataTypes/Request/ApiUpdateUserRequest.cs:21]: i did this [here](https://github.com/LittleBigRefresh/Refresh/pull/520/files#diff-82503085bd955b6efc6eb6353f5f08e8cb21ac466bb03eaad9d5b0e7f3a0eaf8R181)

File: Refresh.GameServer/Database/GameDatabaseContext.Users.cs (hunk 1 above)
Comment (Beyley) [reply Refresh.GameServer/Database/GameDatabaseContext.Users.cs:404]: so basically, in the implementation of the endpoint do `get current settings -> update changed fields -> update whole settings`?

Comment (jvyden) [reply Refresh.GameServer/Database/GameDatabaseContext.Users.cs:404]: Just move the logic from the endpoint to a function in the database that applies based off of the privacy settings object. The implementation itself doesn't need to change, apart from putting everything in a write block I guess.

File: Refresh.GameServer/Database/GameDatabaseContext.Users.cs [hunk 4]
```diff
@@ -380,4 +387,16 @@ public void IncrementUserFilesizeQuota(GameUser user, int amount)
             user.FilesizeQuotaUsage += amount;
         });
     }
+    
+    public void SetPrivacySettings(GameUser user, MetadataEndpoints.PrivacySettings settings)
```
Comment (jvyden) [inline Refresh.GameServer/Database/GameDatabaseContext.Users.cs:391]: Ouch. Yeah, I would extract PrivacySettings to its own class. `SerializedPrivacySettings` somewhere in `Types.UserData`.

Generally speaking the database should never depend on endpoints, and ideally never depend on Bunkum or any services. Reason being I hope to someday move the database schema to its own project in the solution.

**Review by jvyden — CHANGES_REQUESTED** [review]

**Review by jvyden — APPROVED** [review]

**Pull request merged**